library/src/scala/NamedTuple.scala

```diff
@@ -0,0 +1,204 @@
+package scala
+import annotation.experimental
+import compiletime.ops.boolean.*
+
+@experimental
+object NamedTuple:
+
+  /** The type to which named tuples get mapped to. For instance,
+   *      (name: String, age: Int)
+   *  gets mapped to
+   *      NamedTuple[("name", "age"), (String, Int)]
+   */
+  opaque type NamedTuple[N <: Tuple, +V <: Tuple] >: V <: AnyNamedTuple = V
+
+  /** A type which is a supertype of all named tuples */
+  opaque type AnyNamedTuple = Any
+
+  def apply[N <: Tuple, V <: Tuple](x: V): NamedTuple[N, V] = x
+
+  def unapply[N <: Tuple, V <: Tuple](x: NamedTuple[N, V]): Some[V] = Some(x)
+
+  extension [V <: Tuple](x: V)
+    inline def withNames[N <: Tuple]: NamedTuple[N, V] = x
+
+  export NamedTupleDecomposition.{Names, DropNames}
+
+  extension [N <: Tuple, V <: Tuple](x: NamedTuple[N, V])
+
+    /** The underlying tuple without the names */
+    inline def toTuple: V = x
+
+    /** The number of elements in this tuple */
+    inline def size: Tuple.Size[V] = toTuple.size
+
+    // This intentionally works for empty named tuples as well. I think NonEmptyTuple is a dead end
+    // and should be reverted, just like NonEmptyList is also appealing at first, but a bad idea
+    // in the end.
+
+    /** The value (without the name) at index `n` of this tuple */
+    inline def apply(n: Int): Tuple.Elem[V, n.type] =
+      inline toTuple match
+        case tup: NonEmptyTuple => tup(n).asInstanceOf[Tuple.Elem[V, n.type]]
+        case tup => tup.productElement(n).asInstanceOf[Tuple.Elem[V, n.type]]
+
+    /** The first element value of this tuple */
+    inline def head: Tuple.Elem[V, 0] = apply(0)
+
+    /** The tuple consisting of all elements of this tuple except the first one */
+    inline def tail: Tuple.Drop[V, 1] = toTuple.drop(1)
```

Comment:
is it intentional to drop the labels here? the `NamedTuple.Tail` type preserves labels

Reply:
No, that's an oversight. The labels should be kept.